pkg/util/errors.go

```diff
@@ -186,3 +192,145 @@ func HandleStorageError(err error, entityName string) error {
+		return &HTTPError{
+			ErrorType:   "BadRequest",
+			Description: errorMessage,
+			StatusCode:  http.StatusBadRequest,
+		}
+	default:
+		return err
+	}
+}
+
+var (
+	ErrCatalogUsesReservedPlanName           = errors.New("catalog contains a reserved plan name")
+	ErrPlanMustBeBindable                    = errors.New("plan must be bindable")
+	ErrAsyncNotSupportedForSharing           = errors.New("can't use async for sharing instances")
+	ErrReferencedInstanceNotShared           = errors.New("referenced-instance should be shared first")
+	ErrChangingPlanOfReferenceInstance       = errors.New("changing plan of reference instance")
+	ErrChangingPlanOfSharedInstance          = errors.New("changing plan of shared instance")
+	ErrChangingParametersOfReferenceInstance = errors.New("changing parameters of reference instance")
+	ErrMissingOrInvalidReferenceParameter    = errors.New("missing or invalid referenced_instance_id parameter")
+	ErrParsingNewCatalogWithReference        = errors.New("failed generating reference-plan")
+	ErrUnknownOSBMethod                      = errors.New("osb method is unknown")
+	ErrSharedPlanHasReferences               = errors.New("shared plan has references")
+	ErrInstanceIsAlreadyAtDesiredSharedState = errors.New("instance already at the desired shared state")
+	ErrInvalidShareRequest                   = errors.New("invalid share request")
+)
+
+func HandleInstanceSharingError(err error, entityName string) error {
+	if err == nil {
+		return nil
+	}
+
+	if _, ok := err.(*HTTPError); ok {
+		return err
+	}
+
+	switch err {
+	case ErrCatalogUsesReservedPlanName:
+		return &HTTPError{
+			ErrorType:   "BadRequest",
+			Description: fmt.Sprintf("The plan name you used for registration \"%s\" is reserved for the Service Manager; you must choose a different name.", entityName),
+			StatusCode:  http.StatusBadRequest,
+		}
+	case ErrPlanMustBeBindable:
+		return &HTTPError{
+			ErrorType:   "BadRequest",
+			Description: fmt.Sprintf("The plan %s must be defined as 'bindable' so that it can support instance sharing.", entityName),
+			StatusCode:  http.StatusBadRequest,
+		}
+	case ErrReferencedInstanceNotShared:
+		return &HTTPError{
+			ErrorType:   "BadRequest",
+			Description: fmt.Sprintf("Failed to create the reference. The instance %s, for which you want to create the reference, must be shared first.", entityName),
+			StatusCode:  http.StatusBadRequest,
+		}
+	case ErrChangingPlanOfReferenceInstance:
+		return &HTTPError{
+			ErrorType:   "BadRequest",
+			Description: fmt.Sprintf("Failed to update the instance %s. This is a reference instance, therefore its plan can't be changed.", entityName),
+			StatusCode:  http.StatusBadRequest,
+		}
+	case ErrChangingPlanOfSharedInstance:
+		return &HTTPError{
+			ErrorType:   "BadRequest",
+			Description: fmt.Sprintf("Couldn't update the instance's plan. The instance %s is shared, therefore you must unshare it first.", entityName),
```

Comment:
Unsharing is not necessarily the solution...  we should just state that it is not allowed
"It is not possible to change plan of a shred instance"

Reply:
"Couldn't update the instance's plan. It is not possible to change a plan of a shared instance."

Reply:
changed to `"Couldn't update the instance's plan. It is not possible to change a plan of a shared instance."`

Reply:
Why "Couldn't update the instance's plan"?
In an equivalent error we have used ""Failed to update the instance %s. ...". Let's keep consistency.